### PR TITLE
Use Drake CMakeLists.txt and build libbot that is not included in dra…

### PR DESCRIPTION
**Pull request by fbudin69500**
…ke anymore

---
This change is [Reviewable](https://reviewable.io/reviews/robotlocomotion/spartan/228)

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -173,7 +118,7 @@ ExternalProject_Add(director
     -DUSE_PERCEPTION:BOOL=${WITH_PERCEPTION}
     ${CMAKE_ARGS_FOR_EXTERNALS}
   INSTALL_COMMAND ""
-  DEPENDS drake
+  DEPENDS libbot
```
Comment (jamiesnape) [inline CMakeLists.txt]: Still depends on `drake`, I think (`-DUSE_DRAKE:BOOL=ON`).

**Comment by manuelli** [comment]
jenkins please test

**Comment by gizatt** [comment]
jenkins please test

**Comment by manuelli** [comment]
Review status: 0 of 3 files reviewed at latest revision, 1 unresolved discussion.

---

*[CMakeLists.txt, line 115 at r2](https://reviewable.io:443/reviews/robotlocomotion/spartan/228#-L270YuysvBVCX3JeCAv:-L270YuysvBVCX3JeCAw:b9wfpkl) ([raw file](https://github.com/robotlocomotion/spartan/blob/bebe26292519ff88853e59400da5628f8ab4696b/CMakeLists.txt#L115)):*
> ```CMake
>     USES_TERMINAL_BUILD 1
>     CONFIGURE_COMMAND ""
>     BUILD_COMMAND ${BAZEL_ENV} "${BAZEL}" ${BAZEL_DRAKE_STARTUP_ARGS} build ${BAZEL_ARGS} //examples/kuka_iiwa_arm/... //:install @optitrack_driver//src:optitrack_client
> ```

If I am reading this correctly it seems like we are now building drake with CMake. Is this correct? I thought drake only supported Bazel builds, has this changed?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/spartan/228)*

**Comment by jamiesnape** [comment]
Review status: 0 of 3 files reviewed at latest revision, 1 unresolved discussion.

---

*[CMakeLists.txt, line 115 at r2](https://reviewable.io:443/reviews/robotlocomotion/spartan/228#-L270YuysvBVCX3JeCAv:-L270tYEyk-K79mJ0JSR:b-iwdeyd) ([raw file](https://github.com/robotlocomotion/spartan/blob/bebe26292519ff88853e59400da5628f8ab4696b/CMakeLists.txt#L115)):*
<details><summary><i>Previously, manuelli (Lucas Manuelli) wrote…</i></summary><blockquote>

If I am reading this correctly it seems like we are now building drake with CMake. Is this correct? I thought drake only supported Bazel builds, has this changed?
</blockquote></details>

We have abstracted the Bazel build. You still need Bazel to be installed, but otherwise you can just consume Drake as if it were a CMake project.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/spartan/228)*

**Comment by manuelli** [comment]
Review status: 0 of 3 files reviewed at latest revision, 1 unresolved discussion.

---

*[CMakeLists.txt, line 115 at r2](https://reviewable.io:443/reviews/robotlocomotion/spartan/228#-L270YuysvBVCX3JeCAv:-L271Rgpa96bBzd5qp1e:b-a842fr) ([raw file](https://github.com/robotlocomotion/spartan/blob/bebe26292519ff88853e59400da5628f8ab4696b/CMakeLists.txt#L115)):*
<details><summary><i>Previously, jamiesnape (Jamie Snape) wrote…</i></summary><blockquote>

We have abstracted the Bazel build. You still need Bazel to be installed, but otherwise you can just consume Drake as if it were a CMake project.
</blockquote></details>

Ok great. And previously we were doing some magic things inside the docker where we would mount the .bazel-build folder because that's where bazel secretly puts build artifacts. Now I see that the build dir is just `spartan/build/drake`. Is it really here or is it symlinked from the .bazel-build folder? We will need to update procman files to reflect the new location of executables, previously we were looking in `spartan/drake/bazel-build/` for executables.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/spartan/228)*

**Comment by jamiesnape** [comment]
Review status: 0 of 3 files reviewed at latest revision, 1 unresolved discussion.

---

*[CMakeLists.txt, line 115 at r2](https://reviewable.io:443/reviews/robotlocomotion/spartan/228#-L270YuysvBVCX3JeCAv:-L272JADIbx6IZMaHbmp:b-5164vl) ([raw file](https://github.com/robotlocomotion/spartan/blob/bebe26292519ff88853e59400da5628f8ab4696b/CMakeLists.txt#L115)):*
<details><summary><i>Previously, manuelli (Lucas Manuelli) wrote…</i></summary><blockquote>

Ok great. And previously we were doing some magic things inside the docker where we would mount the .bazel-build folder because that's where bazel secretly puts build artifacts. Now I see that the build dir is just `spartan/build/drake`. Is it really here or is it symlinked from the .bazel-build folder? We will need to update procman files to reflect the new location of executables, previously we were looking in `spartan/drake/bazel-build/` for executables.
</blockquote></details>

The executables should all be below either the `bin` or `share/drake` directories in the   install tree now, but let us know if any are missing and we can fix that in drake.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/spartan/228)*

**Comment by manuelli** [comment]
Review status: 0 of 3 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[CMakeLists.txt, line 115 at r2](https://reviewable.io:443/reviews/robotlocomotion/spartan/228#-L270YuysvBVCX3JeCAv:-L2745AMvQg4BYBSef-Z:b-mdb32c) ([raw file](https://github.com/robotlocomotion/spartan/blob/bebe26292519ff88853e59400da5628f8ab4696b/CMakeLists.txt#L115)):*
<details><summary><i>Previously, jamiesnape (Jamie Snape) wrote…</i></summary><blockquote>

The executables should all be below either the `bin` or `share/drake` directories in the   install tree now, but let us know if any are missing and we can fix that in drake.
</blockquote></details>

Ok, could we update the paths in `iiwa_hardware.pmd` to reflect these changes? https://github.com/RobotLocomotion/spartan/blob/master/apps/iiwa/iiwa_hardware.pmd#L8. If not I can commit it as a final fix one you are happy with the changes here.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/spartan/228)*

**Comment by jamiesnape** [comment]
Review status: 0 of 3 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[CMakeLists.txt, line 115 at r2](https://reviewable.io:443/reviews/robotlocomotion/spartan/228#-L270YuysvBVCX3JeCAv:-L275XxoucRS2T0PI_bv:b-hz6v29) ([raw file](https://github.com/robotlocomotion/spartan/blob/bebe26292519ff88853e59400da5628f8ab4696b/CMakeLists.txt#L115)):*
<details><summary><i>Previously, manuelli (Lucas Manuelli) wrote…</i></summary><blockquote>

Do we have an answer on the symlink question?
</blockquote></details>

No symlinks. The artifacts are installed as they would be in a CMake build.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/spartan/228)*

**Comment by manuelli** [comment]
Review status: 0 of 3 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[CMakeLists.txt, line 115 at r2](https://reviewable.io:443/reviews/robotlocomotion/spartan/228#-L270YuysvBVCX3JeCAv:-L276AoPqDkpZRje_ruQ:bn7jn1e) ([raw file](https://github.com/robotlocomotion/spartan/blob/bebe26292519ff88853e59400da5628f8ab4696b/CMakeLists.txt#L115)):*
<details><summary><i>Previously, jamiesnape (Jamie Snape) wrote…</i></summary><blockquote>

No symlinks. The artifacts are installed as they would be in a CMake build.
</blockquote></details>

Is the whole bazel build saved or not. That is my question. Otherwise everytime we quite a docker container then secret .bazel-build folder gets destroyed. So then when we go to build the next time we are essentially having to rebuild all of drake.

So basically does everything gets saved, or only the targets that are installed by drake? Also it would be nice to have access to everything built by drake, in case we need to use something that isn't installed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/spartan/228)*

**Comment by jamiesnape** [comment]
Review status: 0 of 3 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[CMakeLists.txt, line 115 at r2](https://reviewable.io:443/reviews/robotlocomotion/spartan/228#-L270YuysvBVCX3JeCAv:-L276sA5UxP8oIr7iXvQ:b49ba7x) ([raw file](https://github.com/robotlocomotion/spartan/blob/bebe26292519ff88853e59400da5628f8ab4696b/CMakeLists.txt#L115)):*
<details><summary><i>Previously, manuelli (Lucas Manuelli) wrote…</i></summary><blockquote>

Is the whole bazel build saved or not. That is my question. Otherwise everytime we quite a docker container then secret .bazel-build folder gets destroyed. So then when we go to build the next time we are essentially having to rebuild all of drake.

So basically does everything gets saved, or only the targets that are installed by drake? Also it would be nice to have access to everything built by drake, in case we need to use something that isn't installed.
</blockquote></details>

If you want to change the build location you need to set the option in `~/.bazelrc`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/spartan/228)*

**Comment by jamiesnape** [comment]
Review status: 0 of 3 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[CMakeLists.txt, line 115 at r2](https://reviewable.io:443/reviews/robotlocomotion/spartan/228#-L270YuysvBVCX3JeCAv:-L277VErm9WCD-2NH_NJ:b-2dv71t) ([raw file](https://github.com/robotlocomotion/spartan/blob/bebe26292519ff88853e59400da5628f8ab4696b/CMakeLists.txt#L115)):*
<details><summary><i>Previously, jamiesnape (Jamie Snape) wrote…</i></summary><blockquote>

If you want to change the build location you need to set the option in `~/.bazelrc`.
</blockquote></details>

`echo "startup --output_base=/path/to/secret/folder" >> ~/.bazelrc`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/spartan/228)*

**Comment by manuelli** [comment]
Review status: 0 of 3 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[CMakeLists.txt, line 115 at r2](https://reviewable.io:443/reviews/robotlocomotion/spartan/228#-L270YuysvBVCX3JeCAv:-L2ApbRUXbtCnOZy6HxR:b-czoady) ([raw file](https://github.com/robotlocomotion/spartan/blob/bebe26292519ff88853e59400da5628f8ab4696b/CMakeLists.txt#L115)):*
<details><summary><i>Previously, jamiesnape (Jamie Snape) wrote…</i></summary><blockquote>

`echo "startup --output_base=/path/to/secret/folder" >> ~/.bazelrc`
</blockquote></details>

I think I'm not explaining what I mean very well. We had a long discussion about this previously with Jeremy. See https://github.com/RobotLocomotion/spartan/pull/131

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/spartan/228)*

**Comment by jamiesnape** [comment]
Review status: 0 of 3 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[CMakeLists.txt, line 115 at r2](https://reviewable.io:443/reviews/robotlocomotion/spartan/228#-L270YuysvBVCX3JeCAv:-L2DSQuzEJn_hFLgpnZD:b-og3a89) ([raw file](https://github.com/robotlocomotion/spartan/blob/bebe26292519ff88853e59400da5628f8ab4696b/CMakeLists.txt#L115)):*
<details><summary><i>Previously, manuelli (Lucas Manuelli) wrote…</i></summary><blockquote>

I think I'm not explaining what I mean very well. We had a long discussion about this previously with Jeremy. See https://github.com/RobotLocomotion/spartan/pull/131
</blockquote></details>

If you set that in your bazelrc the end result will be identical to what you had before, but you should never have to run anything from the build directory. Everything you need should be installed, and if it is not for some reason you only need to create an issue or a PR and we will add it.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/spartan/228)*

**Comment by gizatt** [comment]
I see what Lucas is getting at -- adding a home directory `.bazelrc` ought to be able to do what we needed in terms of changing the Bazel startup args to put the build artifacts where we want them, so I'm OK with this landing with this feature set. I'm happy to patch back in the functionality we need with some minor tweaks to the Docker setup to get that bazelrc in place.

That said, having the ability to pass arbitrary Bazel startup and regular args (the ones both before and after the "build" or "run" command) through to Drake through a CMake option would be really convenient and help to avoid having to keep track of another configuration file floating around. That's a Drake feature request, anyway.

Btw, the reason we want the build artifacts in Spartan is not so we can use them from that artifact directory (though having them nearby _does_ enable quick hacks to grab uninstalled files before we can make them available through a proper Drake change...) -- rather, because we tend to build in Docker and want those artifacts put somewhere that they get conserved between Docker runs, to avoid having to rebuild Drake from scratch every time we restart the container.

**Comment by jamiesnape** [comment]
> That said, having the ability to pass arbitrary Bazel startup and regular args (the ones both before and after the "build" or "run" command) through to Drake through a CMake option would be really convenient and help to avoid having to keep track of another configuration file floating around. That's a Drake feature request, anyway.

The omission is deliberate. We test in CI every option that we provide for downstream projects. You can achieve everything with a `~/.bazelrc` or per project `.bazelrc` if you really need the functionality.

**Comment by manuelli** [comment]
I have been testing this out but am encountering some build errors. I implemented the `.bazelrc` workaround mentioned but am running into issues building `libbot`. In particular 

```
[ 30%] Performing configure step for 'bot2-vis'
loading initial cache file /home/manuelli/spartan/build/libbot/bot2-vis-prefix/tmp/bot2-vis-cache-Release.cmake
CMake Error: The following variables are used in this project, but they are set to NOTFOUND.
Please set them or make sure they are set and tested correctly in the CMake files:
GLUT_Xmu_LIBRARY (ADVANCED)
    linked by target "bot2-vis" in directory /home/manuelli/spartan/build/externals/libbot/bot2-vis/src/bot_vis
    linked by target "bot-rwx-viewer" in directory /home/manuelli/spartan/build/externals/libbot/bot2-vis/src/rwx-viewer
    linked by target "bot-wavefront-viewer" in directory /home/manuelli/spartan/build/externals/libbot/bot2-vis/src/wavefront-viewer
```

It seems like this requires `libxmu-dev libxi-dev`. I have added them to `install_dependencies.sh.` I opened a new PR with my fixes #233
